### PR TITLE
segregated momentum solves

### PR DESCRIPTION
**Pull request by michaeljbrazell**
- added capability to use segregated momentum/velocity solves (thank you @sayerhs for the hints)
- to use segregated solves set velocity_diffusion.use_tensor_operator = false
- added regression test
- not much difference in speed between tensor and scalar but further testing needed
- the exciting part of this feature is that we can now use NSolve (instead of hypre) which can coarsen overset masking. Preliminary testing shows a 2x speedup for overset sphere case.
- To use NSolve set: 
mac_proj.do_agglomeration=false
mac_proj.do_nsolve = true
diffusion.do_agglomeration=false
diffusion.do_nsolve = true
- segregated solve tested on GABLS case:
velx
vely
[vv.pdf](https://github.com/Exawind/amr-wind/files/6339965/vv.pdf)
[w.pdf](https://github.com/Exawind/amr-wind/files/6339966/w.pdf)
[wtheta.pdf](https://github.com/Exawind/amr-wind/files/6339967/wtheta.pdf)
[ww.pdf](https://github.com/Exawind/amr-wind/files/6339968/ww.pdf)
[theta.pdf](https://github.com/Exawind/amr-wind/files/6339960/theta.pdf)
[u.pdf](https://github.com/Exawind/amr-wind/files/6339961/u.pdf)
[uu.pdf](https://github.com/Exawind/amr-wind/files/6339962/uu.pdf)
[v.pdf](https://github.com/Exawind/amr-wind/files/6339963/v.pdf)

### CODE REVIEW

File: amr-wind/equation_systems/icns/icns_diffusion.H [hunk 1]
```diff
@@ -77,6 +69,216 @@ struct DiffusionOp<ICNS, Scheme>
...
+            mlmg.apply(divtau_comp.vec_ptrs(), vel_comp.vec_ptrs());
+        }
+
+        const auto& repo = this->m_pdefields.repo;
+        const int nlevels = repo.num_active_levels();
+        const auto& density = m_density.state(fstate);
+
+#ifdef _OPENMP
+#pragma omp parallel if (amrex::Gpu::notInLaunchRegion())
+#endif
```
Comment (sayerhs) [inline amr-wind/equation_systems/icns/icns_diffusion.H:152]: We should do this for the `MFIter` loop and not `lev`

File: amr-wind/equation_systems/icns/icns_diffusion.H [hunk 2]
```diff
@@ -77,6 +69,216 @@ struct DiffusionOp<ICNS, Scheme>
...
+                        divtau_arr(i, j, k, 1) *= rhoinv;
+                        divtau_arr(i, j, k, 2) *= rhoinv;
+                    });
+            }
+        }
+    }
+
+    void linsys_solve_impl()
+    {
+        FieldState fstate = FieldState::New;
```
Comment (sayerhs) [inline amr-wind/equation_systems/icns/icns_diffusion.H:173]:
```suggestion
        const FieldState fstate = FieldState::New;
```

File: amr-wind/equation_systems/icns/icns_diffusion.H [hunk 3]
```diff
@@ -77,6 +69,216 @@ struct DiffusionOp<ICNS, Scheme>
...
+                const auto& fld = field(lev).const_array(mfi);
+                const auto& rho = density(lev).const_array(mfi);
+
+                amrex::ParallelFor(
+                    bx, ndim,
+                    [=] AMREX_GPU_DEVICE(int i, int j, int k, int n) noexcept {
+                        rhs_a(i, j, k, n) = rho(i, j, k) * fld(i, j, k, n);
+                    });
+            }
+        }
```
Comment (sayerhs) [inline amr-wind/equation_systems/icns/icns_diffusion.H:201]: Can this be just `lincomb`?

File: amr-wind/equation_systems/icns/icns_diffusion.H [hunk 4]
```diff
@@ -77,6 +69,216 @@ struct DiffusionOp<ICNS, Scheme>
...
+template <typename Scheme>
+struct DiffusionOp<ICNS, Scheme>
+    : public DiffSolverIface<typename ICNS::MLDiffOp>
+{
+    std::unique_ptr<ICNSDiffTensorOp> m_tensor_op;
+    std::unique_ptr<ICNSDiffScalarOp> m_scalar_op;
+
+    static_assert(
+        ICNS::ndim == AMREX_SPACEDIM,
+        "DiffusionOp invoked for non-scalar PDE type");
```
Comment (sayerhs) [inline amr-wind/equation_systems/icns/icns_diffusion.H:245]: Bad copy/paste message.

File: amr-wind/equation_systems/icns/icns_diffusion.H [hunk 5]
```diff
@@ -77,6 +69,216 @@ struct DiffusionOp<ICNS, Scheme>
...
+
+    DiffusionOp(PDEFields& fields, const bool has_overset)
+        : DiffSolverIface<typename ICNS::MLDiffOp>(fields, has_overset)
+    {
+
+        bool use_tensor_op = true;
+        amrex::ParmParse pp("velocity_diffusion");
```
Comment (sayerhs) [inline amr-wind/equation_systems/icns/icns_diffusion.H:252]: Probably should use `fields.field.name() + "_diffusion"`

File: amr-wind/equation_systems/icns/icns_diffusion.H [hunk 6]
```diff
@@ -77,6 +69,216 @@ struct DiffusionOp<ICNS, Scheme>
...
+    std::unique_ptr<ICNSDiffScalarOp> m_scalar_op;
+
+    static_assert(
+        ICNS::ndim == AMREX_SPACEDIM,
+        "DiffusionOp invoked for non-scalar PDE type");
+
+    DiffusionOp(PDEFields& fields, const bool has_overset)
+        : DiffSolverIface<typename ICNS::MLDiffOp>(fields, has_overset)
+    {
+
```
Comment (sayerhs) [inline amr-wind/equation_systems/icns/icns_diffusion.H:250]:
```suggestion
```

File: amr-wind/equation_systems/icns/icns_diffusion.H [hunk 7]
```diff
@@ -77,6 +69,216 @@ struct DiffusionOp<ICNS, Scheme>
     }
 };
 
+class ICNSDiffScalarOp : public DiffSolverIface<amrex::MLABecLaplacian>
```
Comment (sayerhs) [inline amr-wind/equation_systems/icns/icns_diffusion.H:72]: Looking at the implementation of this class, it doesn't make sense for `ICNSDiffScalarOp` to be a subclass of `DiffSolverIface`. It is not using any of the data members and is override almost all methods. I recommend making this a standalone class and just having a clean implementation of `setup_operator`

**Comment by michaeljbrazell** [comment]
@sayerhs thank you for the review, I think I got everything

**Review by sayerhs — APPROVED** [review]

**Pull request merged**